spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/ModuleAggregationUtils.java

```diff
@@ -0,0 +1,152 @@
+		return runAggregated(null, modules, null);
+	}
+
+	/**
+	 * Embeds a group of modules into an existing parent context
+	 *
+	 * @param parentContext the parent context
+	 * @param modules a list of classes, representing root context definitions for modules
+	 * @param args arguments for the modules
+	 */
+	public static void runEmbedded(ConfigurableApplicationContext parentContext,
+			Class<?>[] modules, String[][] args) {
+		SharedChannelRegistry bean = parentContext.getBean(SharedChannelRegistry.class);
+		prepareSharedChannelRegistry(bean, modules);
+		// create child contexts first
+		createChildContexts(parentContext, modules, args);
+	}
+
+	private static ConfigurableApplicationContext createParentContext(String[] args) {
+		SpringApplicationBuilder aggregatorParentConfiguration = new SpringApplicationBuilder();
+		aggregatorParentConfiguration
+				.sources(AggregatorParentConfiguration.class)
+				.web(false)
+				.headless(true)
+				.properties(
+						"spring.jmx.default-domain="
+								+ UUID.randomUUID().toString().replace("-", ""));
+		return aggregatorParentConfiguration.run(args);
+	}
+
+	private static void createChildContexts(ConfigurableApplicationContext parentContext,
+																					Class<?>[] modules, String args[][]) {
+		for (int i = modules.length - 1; i >= 0; i--) {
+			String moduleClassName = modules[i].getName();
+			embedModule(parentContext, getNamespace(moduleClassName, i), modules[i])
+					.run(args != null ? args[i] : new String[0]);
+		}
+	}
+
+	private static String getNamespace(String moduleClassName, int index) {
+		return moduleClassName + "_" + index;
+	}
+
+	private static SpringApplicationBuilder embedModule(
+			ConfigurableApplicationContext applicationContext, String namespace,
+			Class<?> module) {
+		return new SpringApplicationBuilder(module)
+				.web(false)
+				.showBanner(false)
+				.properties(
+						BindableProxyFactory.CHANNEL_NAMESPACE_PROPERTY_NAME + "="
+								+ namespace).registerShutdownHook(false)
+				.parent(applicationContext);
+	}
+
+	private static void prepareSharedChannelRegistry(
+			SharedChannelRegistry sharedChannelRegistry, Class<?>[] modules) {
+		DirectChannel sharedChannel = null;
+		for (int i = 0; i < modules.length; i++) {
+			Class<?> module = modules[i];
+			String moduleClassName = module.getName();
+			if (i > 0) {
+				sharedChannelRegistry.putSharedChannel(getNamespace(moduleClassName, i)
+						+ "." + INPUT_CHANNEL_NAME, sharedChannel);
+			}
+			sharedChannel = new DirectChannel();
+			if (i < modules.length - 1) {
+				sharedChannelRegistry.putSharedChannel(getNamespace(moduleClassName, i)
+						+ "." + OUTPUT_CHANNEL_NAME, sharedChannel);
+			}
+		}
+	}
+
+	/**
+	 * Basic configuration for a parent
+	 */
+	@EnableAutoConfiguration
+	@EnableModule
+	public static class AggregatorParentConfiguration {
+
+		@Bean
+		@ConditionalOnMissingBean(SharedChannelRegistry.class)
+		public SharedChannelRegistry sharedChannelRegistry() {
```

Comment:
might be a crazy idea but could the parent context itself (i.e. BeanFactory) *be* the shared channel registry? It's just a Map after all.... or will the parent also contain other beans?